virtual_rainforest/models/hydrology/hydrology_model.py

```diff
@@ -175,106 +288,343 @@ def update(self, time_index: int) -> None:
+        }
+    )
+
+    return output
+
+
+def calculate_vertical_flow(
+    soil_moisture_residual: DataArray,
+    soil_depth: DataArray,
+    soil_moisture_capacity: Union[float, DataArray] = (
+        HydrologyParameters["soil_moisture_capacity"]
+    ),
+    hydraulic_conductivity: Union[float, DataArray] = (
+        HydrologyParameters["hydraulic_conductivity"]
+    ),
+    hydraulic_gradient: Union[float, DataArray] = (
+        HydrologyParameters["hydraulic_gradient"]
+    ),
+    timestep_conversion_factor: float = (HydrologyParameters["seconds_to_month"]),
+    steepness_water_retention_curve: Union[float, DataArray] = (
+        HydrologyParameters["steepness_water_retention_curve"]
+    ),
+    nonlinearily_parameter: Union[float, DataArray] = (
+        HydrologyParameters["nonlinearily_parameter"]
+    ),
+    meters_to_millimeters: float = HydrologyParameters["meters_to_millimeters"],
+) -> DataArray:
+    """Calculate vertical water flow through soil column.
+
+    To calculate the flow of water through unsaturated soil, this function uses the
+    Richards equation which considers both the hydraulic conductivity and the soil water
+    retention curve to account for the varying moisture content.
+
+    First, the function calculates the effective hydraulic conductivity based on the
+    moisture content using the van Genuchten equation. Then, it applies Darcy's law to
+    calculate the water flow rate considering the effective hydraulic conductivity.
+
+    Args:
+        soil_moisture_residual: residual soil moisture, [relative water content]
+        soil_depths: soil depths = length of the flow path, [m]
+        soil_moisture_capacity: soil moisture capacity, [relative water content]
+        hydraulic_conductivity: hydraulic conductivity of soil, [m/s]
+        hydraulic_gradient: hydraulic gradient (change in hydraulic head) along the flow
+            path, positive values indicate downward flow, [m/m]
+        timestep_conversion_factor: factor to convert flow from m3/s to model time step
+        steepness_water_retention_curve: dimensionless parameter in van Genuchten model
+            that describes the steepness of the water retention curve. Smaller values
+            correspond to steeper curves, indicating that the soil retains water at a
+            higher matric potential. Larger values result in flatter curves, indicating
+            that the soil retains water at a lower matric potential.
+        nonlinearily_parameter: dimensionless parameter in van Genuchten model that
+            describes the degree of nonlinearity of the relationship between the
+            volumetric water content and the soil matric potential.
+
+    Returns:
+        volumetric flow rate of water, [m3/timestep]
+    """
+    # Define the soil water retention curve parameters (van Genuchten model)
+    # TODO abs(hydraulic) gradient support for datatype to allow negative values
+    theta = soil_moisture_capacity - (
```

Comment:
Would it make sense to do this calculation in two steps? As I think it would make the expression a bit more readable